multiqc/modules/nonpareil/nonpareil.py

```diff
@@ -0,0 +1,295 @@
+                "max": 100,
+                "min": 0,
+                "suffix": "%",
+                "scale": "RdYlGn-rev",
+                "format": "{:,.2f}",
+            },
+            "C": {
+                "title": "Coverage",
+                "description": "Dataset coverage",
+                "modify": lambda x: x * 100,
+                "max": 100,
+                "min": 0,
+                "suffix": "%",
+                "scale": "RdYlGn",
+                "format": "{:,.2f}",
+            },
+            "consistent": {
+                "title": "Consistent Data?",
+                "description": "Is the data sufficient for accurate estimation?",
+                "modify": lambda x: "Yes" if x == 1 else "No",
+                "scale": False,
+                "hidden": True,
+            },
+            "star": {
+                "title": "Ideal coverage",
+                "description": "Objective coverage in percentage; i.e., coverage value considered near-complete.",
+                "max": 100,
+                "min": 0,
+                "suffix": "%",
+                "scale": False,
+                "hidden": True,
+            },
+            "LRstar": {
+                "title": f"{config.base_count_prefix} Sequencing effort for ideal coverage",
+                "description": f"Projected sequencing effort for nearly complete coverage ({config.base_count_desc})",
+                "modify": lambda x: x * config.base_count_multiplier,
+                "min": 0,
+                "scale": "RdYlGn-rev",
+                "format": "{:,.2f} " + config.base_count_prefix,
+                "shared_key": "base_count",
+            },
+            "modelR": {
+                "title": "Model correlation",
+                "description": "Pearson R for the estimated model",
+                "max": 1,
+                "min": 0,
+                "scale": "RdYlGn",
+                "format": "{:,.3f}",
+                "hidden": True,
+            },
+            "diversity": {
+                "title": "Diversity",
+                "description": "Dataset Nd index of sequence diversity",
+                "min": 0,
+                "scale": "RdYlGn",
+                "format": "{:,.2f}",
+            },
+        }
+
+        self.general_stats_addcols(self.data_by_sample, headers)
+
+    def nonpareil_redundancy_plot(self):
+        """Make the redundancy plot for nonpareil"""
+
+        esconfig = {
+            "dashStyle": "Dash",
+            "lineWidth": 2,
+            "color": "#000000",
+            "marker": {"enabled": False},
+            "enableMouseTracking": False,
+            "showInLegend": False,
+        }
+        # desc = " **The dashed black line shows theoretical GC content:** `{}`".format(theoretical_gc_name)
+
+        data_plot = list()
+        data_labels = list()
+        extra_series = list()
+        for s_name, data in sorted(self.data_by_sample.items()):
+            data_plot.append(dict())
```

Comment:
Having only one sample per plot is generally something that I try to avoid in MultiQC:

* Doesn't scale well to many samples - tabs are quickly impossible to read / use
* Can't get an overview of all samples in one glance, which is the main purpose of MultiQC

Is there a reason that all plots cannot be shown on a single plot here?

Reply:
I thought about it but, since each sample can have up to 4 curves (depending on user input: observed, model, and dispersion upper and lower), I thought it would be too confusing to have all in the same plot.

I also thought about something similar to FastQC base composition (expandable heatmap) but that is not a standard plot.

Reply:
Can it be four separate plots?

Reply:
You mean 4 plots, each with all the samples? Don't think so, since these plots make most sense together: the observed curve (your data), next to the expected curve under the model and the uncertainty of the model.

I think the closest we can get to having all samples in the same plot, is to do something similar to FastQC base composition:
- heatmap with the (e.g.) observed curve (or, if possible, user selected) for all samples 
- clicking on a row/sample, expands to show the 4 curves for that sample (observed, model and upper/lower dispersion)

However, I looked into it, but I am not sure how to implement this plot since there is a lot of custom code.

@jfy133 since it seems you are also planning to use it, do you have any ideas?

Reply:
Why does it have to be a heatmap? If it's showing a single variable (observed curve) could it not just be a line graph? Click reactivity doesn't depend on it being a heatmap.

Whilst clicking to show more is possible, it should be avoided if we can. It bloats the report size so prevents scaling to large sample numbers and the code is complex and difficult to maintain.

It's usually around now that I say that MultiQC is to summarise data and doesn't have to replicate everything available 😅

Reply:
I've not had a chance to test the current implementation but:

I agree one plot per sample will be overkilll... I'm not sure I completely follow about the heatmap on how/what that is representing, but I agree - you can't really have four plots separate, they are not so informative

For reference, this is the multi-sample PNG that nonpareil produces (that I currently just add to MultiQC as a custom image)

![image](https://github.com/MultiQC/MultiQC/assets/17950287/daf94240-faf0-473a-abdd-1c20989201c6)

and a for a single sample (not from MultiQC ,just default prodcued by the R script): 

![image](https://github.com/MultiQC/MultiQC/assets/17950287/1587d1b3-eef4-4b9c-89f6-b801a89cb30c)

Given MultiQC is meant to be a summary thing only, perhaps just two plots: one of the observed, and a second of  models is  sufficient to help identify outliers? I think getting into dispersion etc. is getting too technical for aMultiQC plot for example

Reply:
Sounds good. Two plots one after another, or one plot with two datasets (two tabbed views)? I'm thinking the latter, as they're essentially showing the same thing?

Reply:
Tabbed would also work IMO, also allows quick switching so should be able to 'see' the observed/model in the same place.

This is just personal opinion though!

Reply:
@jfy133 just to be sure, you are suggesting to have all samples in one plot, but showing only the observed and the model, right?

I can all samples in a single plot and, since the user can choose which plots to show (observed and model are enabled and dispersion is disabled by default), we can leave it up to the user what to show.

Reply:
Yes, I think it would be fine to have all samples in one, and only observed/model as you say (either observed/model is in the same, or across two plots you can switch between I don't mind). 

The dispersion is when you really want to look closely and IMO at that point you would like at the raw data/nonpareil generated plots themselves  etc.

Reply:
Nice 👍🏻  To clarify, I'm talking two datasets, [as described in the docs](https://multiqc.info/docs/development/plots/#switching-datasets-1). So it'll default to showing observed, and the user can flip the toggle switch to see the model curves instead. I think we're on the same page but just trying to be clear. (please don't show multiple lines per sample in a plot at the same time)

Reply:
I think the model curve is more informative when together with the observed, to show how far your data is from the expected model (the same applies to the dispersion). This is how `nonpareil` shows them, even when you have multiple samples.

I agree that several lines for each sample can be confusing and the user can go to the raw data for more info, but I do not see it as a big problem as long as the user can decide what to show (I even think it is nice to give the user control over what to show). What if the report only has one sample and the user wants to see all lines?
Furthermore, the user can also use the side bar to only show the sample of interest for a closer look of the curves. 

So, I still think it is relevant to have a plot with several curves per sample, but it could be an extra dataset. For example, we could have one dataset for the observed, one for the model, and a last one (optional and disabled by default) with all curves.

Reply:
Just pushed some changes. Briefly:
- Removed plotting of dispersion
- Add three tabs: `Observed`, `Model`, `Combined` (both observed and model, with the latter as a dashed line of the same color)